src/media-server.js

```diff
@@ -0,0 +1,148 @@
+import { once } from 'events'
+import fastify from 'fastify'
+import pTimeout from 'p-timeout'
+import StateMachine from 'start-stop-state-machine'
+
+import BlobServerPlugin from './fastify-plugins/blobs.js'
+import { kBlobStore } from './mapeo-project.js'
+
+export const BLOBS_PREFIX = 'blobs'
+export const ICONS_PREFIX = 'icons'
+
+/**
+ * @typedef {Object} StartOpts
+ *
+ * @property {string} [host]
+ * @property {number} [port]
+ */
+
+export class MediaServer {
+  #serverState
+  #fastify
+  #createFastify
+
+  /**
+   * @param {object} params
+   * @param {(projectPublicId: string) => Promise<import('./mapeo-project.js').MapeoProject>} params.getProject
+   * @param {import('fastify').FastifyServerOptions['logger']} [params.logger]
+   */
+  constructor({ getProject, logger }) {
+    this.#createFastify = () => {
+      const server = fastify({ logger })
+
+      server.register(BlobServerPlugin, {
+        prefix: BLOBS_PREFIX,
+        getBlobStore: async (projectPublicId) => {
+          const project = await getProject(projectPublicId)
+          return project[kBlobStore]
+        },
+      })
+
+      return server
+    }
+
+    this.#fastify = this.#createFastify()
+
+    this.#serverState = new StateMachine({
+      start: this.#startServer.bind(this),
+      stop: this.#stopServer.bind(this),
+    })
+  }
+
+  /**
+   * @param {StartOpts} [opts]
+   */
+  async #startServer({ host = '127.0.0.1', port } = {}) {
+    await this.#fastify.listen({ host, port })
+  }
+
+  async #stopServer() {
+    await this.#fastify.close()
+  }
+
+  /**
+   * @returns {Promise<string>}
+   */
+  async #getAddress() {
+    return pTimeout(getServerAddress(this.#fastify.server), {
+      milliseconds: 1000,
+    })
+  }
+
+  /**
+   * @param {StartOpts} [opts]
+   */
+  async start(opts) {
+    // Server was stopped before
+    // Need to replace the existing Fastify instance with a new one in order to listen again
+    if (this.#serverState.state.value === 'stopped') {
+      this.#fastify = this.#createFastify()
+    }
```

Comment:
apparently replacing the fastify instance is the only way to re-start the server? maybe there's something i missed in my search but thought there would be something built-in for this (or a plugin)

Reply:
Best place to search is our own code 🙂 https://github.com/digidem/mapeo-mobile/blob/develop/src/backend/upgrade-manager/upgrade-server.js#L201-L228

Reply:
running into issues attempting to port that implementation. 

I removed the conditional check here, and in `#startServer()` I have:

```ts
if (this.#fastifyStarted) {
  const { server } = this.#fastify
  await promisify(server.listen.bind(server))(this.#port, this.#host)
} else {
  await this.#fastify.listen({ host: this.#host, port: this.#port })
  this.fastifyStarted = true
}
```

running into a type issue with the `promisify` and after a brief search, my guess is that it's maybe related to `@types/node`. even so, running through a lifecycle test, I still get a fastify error about the server already listening (`FastifyError [Error]: Fastify is already listening`). Guessing there's still something I'm missing...

Reply:
i was setting `this.fastifyStarted` instead of `this.#fastifyStarted`... hours wasted on this 🙃

Reply:
re: the promisify type error, seems like it's complaining for a potentially legitimate reason?

seems like the callback param of `server.listen()` doesn't follow the `cb: (err, val) => void` convention. turns out that the callback is only ever called on success, so you'll never get a rejection when wrapping with promisify.

https://nodejs.org/docs/latest-v16.x/api/net.html#serverlisten:

> This function is asynchronous. When the server starts listening, the 'listening' event will be emitted. The last parameter callback will be added as a listener for the 'listening' event.

Seems like doing something along the lines of this would be work though: https://github.com/nodejs/node/issues/21482#issuecomment-626025579